models/path.py

```diff
@@ -38,6 +38,7 @@ def choose_vlans(self, controller):
         for link in self:
             tag_value = link.get_next_available_tag(controller, link.id)
             tag = TAG('vlan', tag_value)
+            link.remove_metadata("s_vlan")
```

Comment:
Example with `RingTopo`:
- Create the EVC:
```
{
    "name": "evc-100", 
    "dynamic_backup_path": true,
    "enabled":true,
    "uni_a":{"interface_id":"00:00:00:00:00:00:00:01:1","tag":{"tag_type":1,"value":100}},
    "uni_z":{"interface_id":"00:00:00:00:00:00:00:02:2","tag":{"tag_type":1,"value":100}}
}
```
- Update with:
```
{
    "dynamic_backup_path": false,
    "primary_path": [{"id": "c8b55359990f89a5849813dc348d30e9e1f991bad1dcb7f82112bd35429d9b07",
      "endpoint_a": {"id": "00:00:00:00:00:00:00:01:4",
       "name": "s1-eth4",
       "port_number": 4,
       "mac": "b2:ff:88:63:33:d8",
       "switch": "00:00:00:00:00:00:00:01",
       "type": "interface",
       "nni": true,
       "uni": false,
       "speed": 1250000000.0,
       "metadata": {},
       "lldp": true,
       "active": true,
       "enabled": true,
       "status": "UP",
       "status_reason": [],
       "link": "c8b55359990f89a5849813dc348d30e9e1f991bad1dcb7f82112bd35429d9b07"},
      "endpoint_b": {"id": "00:00:00:00:00:00:00:03:3",
       "name": "s3-eth3",
       "port_number": 3,
       "mac": "d2:e7:8b:4c:09:4e",
       "switch": "00:00:00:00:00:00:00:03",
       "type": "interface",
       "nni": true,
       "uni": false,
       "speed": 1250000000.0,
       "metadata": {},
       "lldp": true,
       "active": true,
       "enabled": true,
       "status": "UP",
       "status_reason": [],
       "link": "c8b55359990f89a5849813dc348d30e9e1f991bad1dcb7f82112bd35429d9b07"},
      "metadata": {"s_vlan": {"tag_type": "vlan", "value": 999}},
      "active": true,
      "enabled": true,
      "status": "UP",
      "status_reason": []},
     {"id": "4d42dc0852278accac7d9df15418f6d921db160b13d674029a87cef1b5f67f30",
      "endpoint_a": {"id": "00:00:00:00:00:00:00:02:3",
       "name": "s2-eth3",
       "port_number": 3,
       "mac": "56:61:2f:31:93:c3",
       "switch": "00:00:00:00:00:00:00:02",
       "type": "interface",
       "nni": true,
       "uni": false,
       "speed": 1250000000.0,
       "metadata": {},
       "lldp": true,
       "active": true,
       "enabled": true,
       "status": "UP",
       "status_reason": [],
       "link": "4d42dc0852278accac7d9df15418f6d921db160b13d674029a87cef1b5f67f30"},
      "endpoint_b": {"id": "00:00:00:00:00:00:00:03:2",
       "name": "s3-eth2",
       "port_number": 2,
       "mac": "62:b8:88:02:55:e5",
       "switch": "00:00:00:00:00:00:00:03",
       "type": "interface",
       "nni": true,
       "uni": false,
       "speed": 1250000000.0,
       "metadata": {},
       "lldp": true,
       "active": true,
       "enabled": true,
       "status": "UP",
       "status_reason": [],
       "link": "4d42dc0852278accac7d9df15418f6d921db160b13d674029a87cef1b5f67f30"},
      "metadata": {"s_vlan": {"tag_type": "vlan", "value": 999}},
      "active": true,
      "enabled": true,
      "status": "UP",
      "status_reason": []}]
}
```
- The EVC creates a flows with VLANs as 999 but it was not taken from `available_tags`:
```
kytos $> controller.switches["00:00:00:00:00:00:00:03"].interfaces[3].available_tags
Out[1]: {'vlan': [[2, 3798], [3800, 4095]]}

kytos $> controller.switches["00:00:00:00:00:00:00:03"].flows[4].as_dict()
Out[4]: 
{'switch': '00:00:00:00:00:00:00:03',
 'table_id': 0,
 'match': {'in_port': 2, 'dl_vlan': 999},
 'priority': 20000,
 'idle_timeout': 0,
 'hard_timeout': 0,
 'cookie': 12286835438830739018,
 'id': '5e8fc9048dea16f8017b22478e3af674',
 'stats': {'byte_count': 0,
  'duration_sec': 442,
  'duration_nsec': 919000000,
  'packet_count': 0},
 'cookie_mask': 0,
 'instructions': [{'instruction_type': 'apply_actions',
   'actions': [{'vlan_id': 999, 'action_type': 'set_vlan'},
    {'port': 3, 'action_type': 'output'}]}]}
```

Reply:
> isn't this update masking a potential problem that could leak to a leak

Could you explain more about this? `mef_eline` does not reuse paths. If a path is deleted we find another one and choose new VLANs. If there was an error removing flows, `mef_eline` logs the errors and discards the path.

Reply:
> let's plan for an e2e to make sure the problematic situation won't regress too.

Yes, running e2e.

Reply:
@Alopalao, got it. Now it's crystal clear, and once again, good finding. 

The root issue is that the update payload is trusting and accepting link metadata that includes `s_vlan` when using [`_link_from_dict(link_data`)](https://github.com/kytos-ng/mef_eline/blob/master/main.py#L1108), we should only trust the metadata from a link_dict when reading the data from the DB. `"metadata": {"s_vlan": {"tag_type": "vlan", "value": 9999}}` for instance would even lead to other problems since it'd be outside of the valid range. 

So, far we don't allow users to try to specify which s_vlan explicitly they want to use per link, that might be a future feature if one day is needed, but currently they should only set the tag ranges and `mef_eline` application allocates/deallocates accordingly as you know. The only flexibility we should provide when choosing a svlan will be the one you'll contribute on https://github.com/kytos-ng/mef_eline/issues/543.

Can you double check this and see if you share the same understanding? 

If we're on the same page, I believe the fix should only be:

- Validating not to allow update or creation to set link metadata. 
- Consider raising a tag related error if `link.add_metadata(s_tag)` didn't succeed

Reply:
The shape of the link data that's expected is pretty much just the endpoint ids as we have on our e2e tests:

https://github.com/kytos-ng/kytos-end-to-end-tests/blob/master/tests/test_e2e_10_mef_eline.py#L1116-L1124